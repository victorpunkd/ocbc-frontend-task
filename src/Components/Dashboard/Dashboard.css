.dashboaedContainer {
  width: 100%;
}

.dashboardPage {
  height: 75vh;
  overflow: scroll;
  padding: 5% 0%;
}

.dashboardPage::-webkit-scrollbar {
  display: none;
}

.summaryContainer {
  border-radius: 15px;
  background: #642b73; /* fallback for old browsers */
  background: -webkit-linear-gradient(
    to right,
    #c6426e,
    #642b73
  ); /* Chrome 10-25, Safari 5.1-6 */
  background: linear-gradient(
    to right,
    #c6426e,
    #642b73
  ); /* W3C, IE 10+/ Edge, Firefox 16+, Chrome 26+, Opera 12+, Safari 7+ */
  color: white;
  box-shadow: 0px 3px 5px 2px rgba(221, 84, 165, 0.4);
  padding: 5% 5%;
}

.transactionsContainer {
  border-top: 1px solid #c6426e;
  margin-top: 5%;
  padding: 5% 0%;
}

.transactionTableHeader {
  margin-top: 5%;
  display: flex;
}

.headerText {
  flex: 1;
  font-size: 20px;
  font-weight: 700;
  color: gray;
}

.footer {
}
