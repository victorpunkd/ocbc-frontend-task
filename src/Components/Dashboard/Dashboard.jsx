import React, { useCallback, useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import "./Dashboard.css";
import NavBar from "../NavBar/NavBar";
import CustomSnackBar from "../../CommonComponents/CustomSnackBar/CustomSnackBar";
import CustomButton from "../../CommonComponents/CustomButton/CustomButton";
import TransactionLine from "../TransactionLine/TransactionLine";
import CustomBackdrop from "../../CommonComponents/CustomBackdrop/CustomBackdrop";
import axios from "axios";
import CompareArrowsIcon from "@mui/icons-material/CompareArrows";

const Dashboard = () => {
  const navigate = useNavigate();
  const [isBackdropOpen, setIsBackdropOpen] = useState(false);
  const [balance, setBalance] = useState(0);
  const [transactions, setTransactions] = useState([]);
  const [isSnackbarOpen, setIsSnackbarOpen] = useState(false);
  const [snackBarSeverity, setSnackBarSeverity] = useState("");
  const [snackBarMessage, setSnackBarMessage] = useState("");

  const handleMakeTransferClick = () => {
    navigate("/make-transfer");
  };

  const handleBackEndError = useCallback(
    (error) => {
      setIsSnackbarOpen(true);
      setSnackBarMessage(error?.response?.data?.description || "");
      setSnackBarSeverity("error");
      sessionStorage.removeItem("token");
      setTimeout(() => {
        navigate("/");
      }, 1000);
    },
    [navigate]
  );

  const getBalanceCallback = useCallback(() => {
    setIsBackdropOpen(true);
    axios
      .get(`${process.env.REACT_APP_BASE_URL}account/balances`, {
        headers: {
          Authorization: sessionStorage.getItem("token"),
        },
      })
      .then((response) => {
        if (response.data && response.data.balance) {
          setBalance(response.data.balance);
        }
      })
      .catch((error) => {
        if (error.response.data && error.response.data.status) {
          handleBackEndError(error);
        }
      })
      .finally(() => {
        setIsBackdropOpen(false);
      });
  }, [handleBackEndError]);

  const getTransactionsCallback = useCallback(() => {
    setIsBackdropOpen(true);
    axios
      .get(`${process.env.REACT_APP_BASE_URL}account/transactions`, {
        headers: {
          Authorization: sessionStorage.getItem("token"),
        },
      })
      .then((response) => {
        if (response.data && response.data.data) {
          setTransactions(response.data.data);
        }
      })
      .catch((error) => {
        if (error.response.data && error.response.data.status) {
          handleBackEndError(error);
        }
      })
      .finally(() => {
        setIsBackdropOpen(false);
      });
  }, [handleBackEndError]);

  useEffect(() => {
    if (!sessionStorage.getItem("token")) {
      navigate("/no-access");
    } else {
      getBalanceCallback();
      getTransactionsCallback();
    }
  }, [getBalanceCallback, getTransactionsCallback, navigate]);

  return (
    <div className="dashboaedContainer">
      <NavBar />
      <CustomBackdrop
        open={isBackdropOpen}
        handleClose={() => setIsBackdropOpen(false)}
      />
      <CustomSnackBar
        open={isSnackbarOpen}
        severity={snackBarSeverity}
        message={snackBarMessage}
        handleClose={() => setIsSnackbarOpen(false)}
      />
      <div className="dashboardPage">
        <div className="summaryContainer">
          <div style={{ fontWeight: 500 }}>You have</div>
          <div style={{ marginTop: "5%", fontSize: "25px" }}>
            SGD{" "}
            <span style={{ fontWeight: 700 }}>
              {balance &&
                balance.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ",")}
            </span>
          </div>
          <div style={{ fontWeight: 500, marginTop: "5%" }}>
            In your account
          </div>
        </div>
        <div className="transactionsContainer">
          <div style={{ fontWeight: 700, color: "#dd54a5" }}>Your activity</div>
          <div className="transactionTableHeader">
            <div className="headerText">Date</div>
            <div className="headerText">Description</div>
            <div className="headerText" style={{ textAlign: "right" }}>
              Amount
            </div>
          </div>
          {transactions.map((data) => (
            <TransactionLine
              date={data.date}
              description={data.description}
              amount={data.amount}
              transactionType={data.type}
            />
          ))}
        </div>
      </div>
      <div className="footer">
        <CustomButton
          id="makeTransferButton"
          buttonText="Make a transfer"
          buttonClick={handleMakeTransferClick}
          icon={<CompareArrowsIcon />}
        />
      </div>
    </div>
  );
};

export default Dashboard;
