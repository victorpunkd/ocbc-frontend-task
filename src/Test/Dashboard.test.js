import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import Dashboard from "../Components/Dashboard/Dashboard";

configure({ adapter: new Adapter() });

const mockedUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUsedNavigate,
}));

describe("Dashboard", () => {
  it("snapshot renders", () => {
    const component = renderer.create(<Dashboard />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("Should call navigate when Make Transaction button is clicked", () => {
    const wrapper = shallow(<Dashboard />);
    wrapper.find("#makeTransferButton").props().buttonClick();
    expect(mockedUsedNavigate).toBeCalled();
  });
});
