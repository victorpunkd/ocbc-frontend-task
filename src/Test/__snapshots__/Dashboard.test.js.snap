// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard snapshot renders 1`] = `
<div
  className="dashboaedContainer"
>
  <div
    className="navBarContainer"
  >
    <div
      className="logoutButton"
      onClick={[Function]}
    >
      Logout
    </div>
  </div>
  <div>
    <div
      aria-hidden={true}
      className="MuiBackdrop-root css-1ji7omv-MuiBackdrop-root"
      onClick={[Function]}
      style={
        Object {
          "opacity": 0,
          "visibility": "hidden",
        }
      }
    >
      <span
        className="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorInherit css-62e83j-MuiCircularProgress-root"
        role="progressbar"
        style={
          Object {
            "height": 40,
            "width": 40,
          }
        }
      >
        <svg
          className="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx={44}
            cy={44}
            fill="none"
            r={20.2}
            strokeWidth={3.6}
            style={Object {}}
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    className="dashboardPage"
  >
    <div
      className="summaryContainer"
    >
      <div
        style={
          Object {
            "fontWeight": 500,
          }
        }
      >
        You have
      </div>
      <div
        style={
          Object {
            "fontSize": "25px",
            "marginTop": "5%",
          }
        }
      >
        SGD
         
        <span
          style={
            Object {
              "fontWeight": 700,
            }
          }
        >
          0
        </span>
      </div>
      <div
        style={
          Object {
            "fontWeight": 500,
            "marginTop": "5%",
          }
        }
      >
        In your account
      </div>
    </div>
    <div
      className="transactionsContainer"
    >
      <div
        style={
          Object {
            "color": "#dd54a5",
            "fontWeight": 700,
          }
        }
      >
        Your activity
      </div>
      <div
        className="transactionTableHeader"
      >
        <div
          className="headerText"
        >
          Date
        </div>
        <div
          className="headerText"
        >
          Description
        </div>
        <div
          className="headerText"
          style={
            Object {
              "textAlign": "right",
            }
          }
        >
          Amount
        </div>
      </div>
    </div>
  </div>
  <div
    className="footer"
  >
    <button
      className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root makeStyles-buttonStyle-1 css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      id="makeTransferButton"
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Make a transfer
      <span
        className="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CompareArrowsIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M9.01 14H2v2h7.01v3L13 15l-3.99-4v3zm5.98-1v-3H22V8h-7.01V5L11 9l3.99 4z"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
